### PR TITLE
Fix misleading error message in misbehavior detection task

### PR DESCRIPTION
**Pull request by adizere**
Closes: #2031 

## Reproduce & test

To reproduce the problem & test this PR, proceed as follows:

### Setup phase

1. Use this `gm` config to setup 2 chains. Modify the `binary` option to match your setup.

```toml

[global]
  add_to_hermes = false
  auto_maintain_config = true
  gaiad_binary = "$GOPATH/bin/gaiad"
  hdpath = ""
  home_dir = "$HOME/.gm"
  ports_start_at = 27040
  validator_mnemonic = ""
  wallet_mnemonic = ""
  extra_wallets = 1

  [global.hermes]
    binary = "/Users/adi/Hammers/ibc-rs/target/debug/hermes"
    config = "$HOME/.hermes/config.toml"
    log_level = "info"
    strategy = "packets"
    telemetry_enabled = true
    telemetry_host = "127.0.0.1"
    telemetry_port = 3001

[ibc-0]
  ports_start_at = 27000

[ibc-1]
  ports_start_at = 27010

[node0]
  add_to_hermes = true
  network = "ibc-0"
  ports_start_at = 27020

[node1]
  add_to_hermes = true
  network = "ibc-1"
  ports_start_at = 27030
```

2. Make sure you have no `gaia` process running (if you do, kill them). 
3. Now `gm reset` and `gm start`.
4. `gm status` should show something like this:
> $ gm status
NODE               PID    RPC   APP  GRPC  HOME_DIR
ibc-0            89448  27000 27001 27002  /Users/adi/.gm/ibc-0
 node0           89817  27020 27021 27022  /Users/adi/.gm/node0
ibc-1            89606  27010 27011 27012  /Users/adi/.gm/ibc-1
 node1           90028  27030 27031 27032  /Users/adi/.gm/node1

5. `gm hermes config` and `gm hermes keys`
6. edit the default Hermes config file at "$HOME/.hermes/config.toml"
    - replace `trusting_period = '14days'` with `trusting_period = '40s'` for chain `ibc-1`

### Reproduce & test the issue

7. `hermes create channel ibc-0 --port-a transfer --port-b transfer --new-client-connection -c ibc-1`
    - this should create a new channel/conn/client
    - the client hosted on `ibc-0` should have a very brief trusting period, namely `50` seconds

8. __immediately do__ `hermes start` in a terminal
    - this will keep the newly created client refreshed, prevent it from expiring

__Running on master__

9. When you run `start` from master branch, you will see this misleading message
> 2022-04-28T07:57:51.500423Z  WARN ThreadId(29) DetectMisbehaviorWorker{client=07-tendermint-3 src_chain=ibc-1 dst_chain=ibc-0}: [ibc-1 -> ibc-0:07-tendermint-3] misbehaviour checking result: error raised while checking for misbehaviour evidence: failed to check misbehaviour for 07-tendermint-3 at consensus height 1-177: light client error for RPC address ibc-1: trusted state outside of trusting period

The above the is message we're trying to adjust.

__Running from this branch__

Note: You might want to redo steps 7 & 8 after checking out this branch.

9. Adjust the `log_info` in "$HOME/.hermes/config.toml" to be `log_level = 'debug'`.
10. Running `start` from this branch will not show the "DetectMisbehaviorWorker" span any longer, because the error is no longer propagated up the stack. Instead, we can see this 

> 2022-04-28T08:19:24.026431Z DEBUG ThreadId(26) light_client::check_misbehaviour{client_id=07-tendermint-3}: trusted consensus state is outside of trusting period; check_misbehaviour finished OK trusted=1-413 target=1-418

______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules). 
- [x] Linked to GitHub issue.
- [x] Updated code comments and documentation (e.g., `docs/`).

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: relayer/src/foreign_client.rs [hunk 1]
```diff
@@ -1226,10 +1226,6 @@ impl<DstChain: ChainHandle, SrcChain: ChainHandle> ForeignClient<DstChain, SrcCh
             // Check for misbehaviour according to the specific source chain type.
             // In case of Tendermint client, this will also check the BFT time violation if
             // a header for the event height cannot be retrieved from the witness.
-            //
-            // FIXME: This returns error if the update event contains expired client state.
-            // This can happen even if the latest client state is unexpired, but the
-            // update event is from earlier.
             let misbehavior = self
```
Comment (adizere) [inline relayer/src/foreign_client.rs:1229]: Note to reviewers: this is where the `Err` leaked out of `check_misbehaviour` when the trusted state was expired;  we were interpreting that as an abnormal failure in misbehavior check, but it was not a failure. I changed `check_misbehaviour` to return `Ok(None)` now, so I also removed the FIXME.

File: relayer/src/light_client/tendermint.rs [hunk 2]
```diff
@@ -141,8 +147,33 @@ impl super::LightClient<CosmosSdkChain> for LightClient {
...
+            target_height,
+            client_state,
+        ) {
+            // Verification passed, return the target block + supporting blocks.
+            Ok(v) => v,
+
+            // Special-case: Verification failed because the trusted state is too old.
+            // This means we cannot check for misbehavior with the provided trusted state.
+            Err(e)
+                if e.trace()
```
[the remark this answers is not in the document]
Comment (soareschen) [reply relayer/src/light_client/tendermint.rs]: I pushed [ce3332f](https://github.com/informalsystems/ibc-rs/pull/2156/commits/ce3332f9dd7791fdd0a90caa05997cfbaacea5c5) to fix that.

[the remark this answers is not in the document]
Comment (adizere) [reply relayer/src/light_client/tendermint.rs]: Thanks! Appreciate it.

[the remark this answers is not in the document]
Comment (adizere) [reply relayer/src/light_client/tendermint.rs]: Feel free to drop a review & test when you get a chance.

File: relayer/src/foreign_client.rs (hunk 1 above)
Comment (ancazamfir) [reply relayer/src/foreign_client.rs:1229]: I wouldn't say it was an "abnormal failure", it was a warning. imo we should still return warn and not Ok. Maybe make the message more clear to indicate that we are stopping misbehaviour due to expired states. Or return Ok but bail out if misbehavior runs for all states is in progress. The problem in this branch is that at start we now continue to go back through all expired consensus states. Recent gaia versions use SDK versions that prune expired consensus states so it is not an issue with these. For gaia v5.0.x this problem shows. 

Also the current debug is a bit confusing, why would misbehavior be Ok with expired states?

Comment (ancazamfir) [reply relayer/src/foreign_client.rs:1229]: Also not sure why this was marked as `FIXME`?

Comment (adizere) [reply relayer/src/foreign_client.rs]: Should be fixed now, thanks for catching that Anca! https://github.com/informalsystems/ibc-rs/pull/2156/commits/ba9831f834403071152ca1c9273a6e31aeba74e2 Please weigh in.

**Review by ancazamfir — APPROVED** [review]
lgtm! thanks @adizere!

**Pull request merged**